graphbolt/src/cuda/unique_and_compact_impl.cu

```diff
@@ -0,0 +1,161 @@
+              sorted_unique_dst_ids.get(), unique_dst_ids.size(0), 0, num_bits,
+              stream));
+          auto temp = allocator.AllocateStorage<char>(workspace_size);
+          CUDA_CALL(cub::DeviceRadixSort::SortKeys(
+              temp.get(), workspace_size, unique_dst_ids_ptr,
+              sorted_unique_dst_ids.get(), unique_dst_ids.size(0), 0, num_bits,
+              stream));
+        }
+        auto is_dst = allocator.AllocateStorage<bool>(src_ids.size(0));
+        thrust::binary_search(
+            exec_policy, sorted_unique_dst_ids.get(),
+            sorted_unique_dst_ids.get() + unique_dst_ids.size(0), src_ids_ptr,
+            src_ids_ptr + src_ids.size(0), is_dst.get());
+        auto only_src = allocator.AllocateStorage<scalar_t>(src_ids.size(0));
+        auto only_src_size =
+            thrust::remove_copy_if(
+                exec_policy, src_ids_ptr, src_ids_ptr + src_ids.size(0),
+                is_dst.get(), only_src.get(), thrust::identity<bool>{}) -
+            only_src.get();
+        auto sorted_only_src =
+            allocator.AllocateStorage<scalar_t>(only_src_size);
+        {
+          size_t workspace_size;
+          CUDA_CALL(cub::DeviceRadixSort::SortKeys(
+              nullptr, workspace_size, only_src.get(), sorted_only_src.get(),
+              only_src_size, 0, num_bits, stream));
+          auto temp = allocator.AllocateStorage<char>(workspace_size);
+          CUDA_CALL(cub::DeviceRadixSort::SortKeys(
+              temp.get(), workspace_size, only_src.get(), sorted_only_src.get(),
+              only_src_size, 0, num_bits, stream));
+        }
+        auto unique_only_src = torch::empty(only_src_size, src_ids.options());
+        auto unique_only_src_ptr = unique_only_src.data_ptr<scalar_t>();
+        auto unique_only_src_cnt = allocator.AllocateStorage<scalar_t>(1);
+        {
+          size_t workspace_size;
+          CUDA_CALL(cub::DeviceRunLengthEncode::Encode(
+              nullptr, workspace_size, sorted_only_src.get(),
+              unique_only_src_ptr, cub::DiscardOutputIterator{},
+              unique_only_src_cnt.get(), only_src_size, stream));
+          auto temp = allocator.AllocateStorage<char>(workspace_size);
+          CUDA_CALL(cub::DeviceRunLengthEncode::Encode(
+              temp.get(), workspace_size, sorted_only_src.get(),
+              unique_only_src_ptr, cub::DiscardOutputIterator{},
+              unique_only_src_cnt.get(), only_src_size, stream));
+        }
+        scalar_t unique_only_src_size;
+        CUDA_CALL(cudaMemcpyAsync(
+            &unique_only_src_size, unique_only_src_cnt.get(),
+            sizeof(unique_only_src_size), cudaMemcpyDeviceToHost, stream));
+        stream.synchronize();
+        unique_only_src = unique_only_src.slice(0, 0, unique_only_src_size);
+        auto real_order = torch::cat({unique_dst_ids, unique_only_src});
+        auto [sorted_order, new_ids] = Sort(real_order, num_bits);
```

Comment:
Why do we need to sort the order in the UniqueAndCompact?

Reply:
added explanations, could you rereview the code to check if all concerns are addressed?

Reply:
All the lookup logic in this implementation is done via first sorting and then binary search. That is why sorting and binary search and lower_bound are used extensively.